cmd/internal/planetscale_edge_database.go

```diff
@@ -285,49 +279,92 @@ func (p PlanetScaleEdgeDatabase) sync(ctx context.Context, tc *psdbconnect.Table
 	for {
 		res, err := c.Recv()
 		if err != nil {
 			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sExiting sync and flushing records due to error: %+v", preamble, err))
 			return tc, resultCount, err
 		}
 
-		if res.Cursor != nil {
-			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sAdvancing cursor to %+v", preamble, res.Cursor))
-			tc = res.Cursor
+		var rows []*query.QueryResult
+		for _, event := range res.Events {
+			switch event.Type {
+			case binlogdata.VEventType_VGTID:
+				vgtid := event.GetVgtid().ShardGtids[0]
+				if vgtid != nil {
+					// Update cursor to new VGTID
+					tc = &psdbconnect.TableCursor{
+						Shard:    tc.Shard,
+						Keyspace: tc.Keyspace,
+						Position: vgtid.Gtid,
+					}
+				}
+			case binlogdata.VEventType_LASTPK:
+				if event.LastPKEvent.TableLastPK != nil {
+					// Only update last PK because we're in a COPY phase
+					tc = &psdbconnect.TableCursor{
+						Shard:       tc.Shard,
+						Keyspace:    tc.Keyspace,
+						LastKnownPk: event.LastPKEvent.TableLastPK.Lastpk,
+						Position:    tc.Position,
+					}
+					p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sLASTPK event found, setting last PK to %+v", preamble, tc))
+				}
+			case binlogdata.VEventType_FIELD:
+				// Save fields for processing
+				fields = event.FieldEvent.Fields
+				p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sFIELD event found, setting fields to %+v", preamble, fields))
+			case binlogdata.VEventType_ROW:
+				// Collect rows for processing
+				for _, change := range event.RowEvent.RowChanges {
+					if change.After != nil {
+						rows = append(rows, &query.QueryResult{
+							Fields: fields,
+							Rows:   []*query.Row{change.After},
+						})
+					}
+				}
+			case binlogdata.VEventType_COPY_COMPLETED:
+				p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sCOPY_COMPLETED event found, copy phase finished", preamble))
+				copyCompletedSeen = true
+			}
 		}
 
-		// Heartbeats and other non-DML queries can create binlog events with the same VGTID, but no rows.
-		// These no-row results can have the same VGTID as a subsequent result with rows.
-		if len(res.Result) > 0 {
-			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sFound %+v results", preamble, len(res.Result)))
-			// Watch for VGTID change as soon as we encounter records from some VGTID that is equal to, or after the stop position we're looking for.
-			// We watch for a VGTID that is equal to or after (not just equal to) the stop position, because by the time the first sync for records occurs,
-			// the current VGTID may have already advanced past the stop position.
-			watchForVgGtidChange = watchForVgGtidChange || positionAtLeast(tc.Position, stopPosition)
-			for _, result := range res.Result {
+		if waitForCopyCompleted && copyCompletedSeen {
+			p.Logger.Log(LOGLEVEL_INFO, fmt.Sprintf("%sCopy phase completed. Waiting for next VGTID after stop position.", preamble))
+			watchForVgGtidChange = true
+		}
+		if !waitForCopyCompleted && positionEqual(tc.Position, stopPosition) {
```

Comment:
During an incremental sync, we'll have one `VStream` response for every VGTID from the start position to the stop position, so it's safe to look for the exact stop position.

Reply:
> so it's safe to stop looking for rows at the exact stop position.
does setting `watchForVgGtidChange = true` make it stop looking for exact stop position?